client/web-sveltekit/src/lib/codenav/ExplorePanel.svelte

```diff
@@ -24,95 +24,150 @@
+            for (const [index, dir] of dirs.entries()) {
+                if (!current.tree.get(dir)) {
+                    current.tree.set(dir, {
+                        tree: new Map(),
+                        entry: {
+                            type: 'dir',
+                            repo,
+                            name: dir,
+                            path: pathElements.slice(0, index + 1).join('/') + '/',
+                        },
+                    })
+                }
+                current = current.tree.get(dir)!
             }
 
-            pathGroups[seenPaths[path]].usages.push(usage)
+            const fileName = pathElements.at(-1)! // splitting will always have at least one element
+            current.tree.set(fileName, {
+                tree: new Map(),
+                entry: {
+                    type: 'file',
+                    repo,
+                    path,
+                    name: fileName,
+                },
+            })
         }
 
-        return repoGroups
-    }
+        for (const pathGroup of pathGroups) {
+            addToTree(pathGroup.repository, pathGroup.path)
+        }
 
-    function treeProviderForEntries(entries: TreeEntry[]): TreeProvider<TreeEntry> {
-        return {
-            getNodeID(entry) {
-                if (entry.type === 'repo') {
-                    return `repo-${entry.name}`
-                } else {
-                    return `path-${entry.repo}-${entry.name}`
-                }
-            },
-            getEntries(): TreeEntry[] {
-                return entries
-            },
-            isExpandable(entry) {
-                return entry.type === 'repo'
-            },
-            isSelectable() {
-                return true
-            },
-            fetchChildren(entry) {
-                if (entry.type === 'repo') {
-                    return Promise.resolve(treeProviderForEntries(entry.entries))
-                } else {
-                    throw new Error('path nodes are not expandable')
-                }
-            },
+        function newTreeProvider(tree: Tree): TreeProvider<TreeEntry> {
+            return {
+                getNodeID(entry) {
+                    if (entry.type === 'repo') {
+                        return `repo-${entry.name}`
+                    } else {
+                        return `path-${entry.repo}-${entry.path}`
+                    }
+                },
+                getEntries(): TreeEntry[] {
+                    return Array.from(tree.entries())
+                        .map(([_name, entry]) => entry.entry)
+                        .toSorted((a, b) => {
+                            if (a.type !== b.type) {
+                                const typeRanks = { repo: 0, dir: 1, file: 2 }
+                                return typeRanks[a.type] < typeRanks[b.type] ? -1 : 1
+                            }
+                            return a.name < b.name ? -1 : 1
```

Comment:
I've run into issues in the past when the compare function didn't return `0` entries that are actually equal. I suggest to do

```suggestion
                            return a.name.localeCompare(b.name)
```

Reply:
TIL, thanks!